### PR TITLE
#4128 Add except as a dynamic policy enforcement point (4-2-stable)

### PR DESCRIPTION
**Pull request by jasoncoposky**

### CODE REVIEW

File: scripts/irods/test/test_native_rule_engine_plugin.py [hunk 1]
```diff
@@ -76,6 +76,75 @@ def helper_test_pep(self, rules_to_add, icommand, strings_to_check_for=['THIS IS
...
+       self.helper_test_pep(pre_pep_fail, 'iget -f '+self.testfile, ['EXCEPT FOR PRE PEP FAIL'])
+
+       op_fail = """
+           pep_resource_open_except(*INST, *CTX, *OUT) {
+               writeLine("serverLog", "EXCEPT FOR OPERATION FAIL")
+           }
+       """
+       self.helper_test_pep(op_fail, 'iget -f bad_file', ['EXCEPT FOR OPERATION FAIL'])
+
+       post_fail = """
```
Comment (trel) [inline scripts/irods/test/test_native_rule_engine_plugin.py]: post_pep_fail

File: scripts/irods/test/test_native_rule_engine_plugin.py [hunk 2]
```diff
@@ -76,6 +76,75 @@ def helper_test_pep(self, rules_to_add, icommand, strings_to_check_for=['THIS IS
...
+           pep_resource_open_post(*INST, *CTX, *OUT) {
+               failmsg(-1, "POST PEP FAIL")
+           }
+           pep_resource_open_except(*INST, *CTX, *OUT) {
+               writeLine("serverLog", "EXCEPT FOR POST PEP FAIL")
+           }
+       """
+       self.helper_test_pep(post_fail, 'iget -f '+self.testfile, ['EXCEPT FOR POST PEP FAIL'])
```
Comment (trel) [inline scripts/irods/test/test_native_rule_engine_plugin.py]: post_pep_fail

File: scripts/irods/test/test_native_rule_engine_plugin.py [hunk 3]
```diff
@@ -76,6 +76,75 @@ def helper_test_pep(self, rules_to_add, icommand, strings_to_check_for=['THIS IS
...
+       self.helper_test_pep(pre_pep_fail, 'iget -f '+self.testfile, ['EXCEPT FOR PRE PEP FAIL'])
+
+       op_fail = """
+           pep_api_data_obj_get_except(*INST, *COMM, *INP, *PORT, *BUF) {
+               writeLine("serverLog", "EXCEPT FOR OPERATION FAIL")
+           }
+       """
+       self.helper_test_pep(op_fail, 'iget -f bad_file', ['EXCEPT FOR OPERATION FAIL'])
+
+       post_fail = """
```
Comment (trel) [inline scripts/irods/test/test_native_rule_engine_plugin.py]: post_pep_fail

File: scripts/irods/test/test_native_rule_engine_plugin.py [hunk 4]
```diff
@@ -76,6 +76,75 @@ def helper_test_pep(self, rules_to_add, icommand, strings_to_check_for=['THIS IS
...
+           pep_api_data_obj_get_post(*INST, *COMM, *INP, *PORT, *BUF) {
+               failmsg(-1, "POST PEP FAIL")
+           }
+           pep_api_data_obj_get_except(*INST, *COMM, *INP, *PORT, *BUF) {
+               writeLine("serverLog", "EXCEPT FOR POST PEP FAIL")
+           }
+       """
+       self.helper_test_pep(post_fail, 'iget -f '+self.testfile, ['EXCEPT FOR POST PEP FAIL'])
```
Comment (trel) [inline scripts/irods/test/test_native_rule_engine_plugin.py]: post_pep_fail

File: lib/core/include/apiHandler.hpp [hunk 5]
```diff
@@ -143,56 +142,70 @@ namespace irods {
...
+                        if(!pre_err.ok()) {
+                            // if the pre-pep fails, invoke the exception pep
+                            error except_err = invoke_policy_enforcement_point(
+                                               re_ctx_mgr,
+                                               ctx,
+                                               operation_name,
+                                               "except",
+                                               std::forward<types_t>(_t)...);
+                            if(!except_err.ok()) {
+                                irods::log(PASS(except_err));
```
Comment (alanking) [inline lib/core/include/apiHandler.hpp]: Did we want to return except_err in this case, or is it more important that the caller knows which PEP initially failed and why? Same for 179 and 200, and in the plugin_base.

Also, would it be worth capturing "pre", "post", and "except" in consts?

**Comment by trel** [comment]
update the commit message (add #) and we'll merge

**Pull request merged**